File: Two pointers/leetcode_16.py
# Problem Link ====>>https://leetcode.com/problems/3sum-closest/description/

class List:
    def __init__(self):
        self.data = []

    def append(self, item):
        self.data.append(item)
class Solution:
    def threeSumClosest(self, nums: List[int], target:int) ->int:
        N = len(nums)

        if N ==3:
            return sum(nums)

        nums = sorted(nums)
        minSum, minDiff = None, int

        for i in range(N - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, N - 1

            while left < right:
                currSum = sum([nums[i], nums[left], nums[right]])
                currDiff = abs(target - currSum)

                if currDiff < minDiff:
                    minDiff = currDiff
                    minSum = currSum

                    if minDiff == 0:
                        return minSum

                if target > currSum:
                    left += 1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                else:
                    right -= 1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1

        return minSum

File: Two pointers/leetcode_16.cpp
// Given an integer array nums of length n and an integer target, find three integers in nums such that the sum is closest to target.
// Return the sum of the three integers.You may assume that each input would have exactly one solution.

#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>
using namespace std;

class Solution
{
public:
    int threeSumClosest(vector<int> &nums, int target)
    {
        int n = nums.size(), sum = 0, ans = 0;
        int diff = INT_MAX;
        sort(nums.begin(), nums.end());

        if (n == 3)
            return nums[0] + nums[1] + nums[2];
        for (int i = 0; i < n; i++)
        {
            int first = nums[i];
            int high = i + 1, low = n - 1;
            while (high < low)
            {
                sum = first + nums[high] + nums[low];
                if (sum == target)
                    return target;
                else if (abs(target - sum) < diff)
                {
                    diff = abs(target - sum);
                    ans = sum;
                }
                if (sum > target)
                    low--;
                else
                    high++;
            }
        }
        return ans;
    }
};

int main()
{
    // Create a vector of integers for nums
    vector<int> nums = {-1, 2, 1, -4};
    int target = 1;

    // Create a Solution object
    Solution solution;

    // Find the sum of three integers in nums that is closest to the target
    int closest_sum = solution.threeSumClosest(nums, target);

    // Print the closest sum value
    cout << "Closest sum to " << target << " in nums: " << closest_sum << endl;

    return 0;
}